src/bin/sccache-dist/main.rs

```diff
@@ -99,72 +201,188 @@ fn main() {
-            let toolchain_cache_size = value_t_or_exit!(matches.value_of("toolchain-cache-size"), u64);
-            let scheduler_addr = value_t_or_exit!(matches.value_of("scheduler-addr"), IpAddr);
-            Command::Server { builder, cache_dir, toolchain_cache_size, scheduler_addr }
+            let config_path = Path::new(matches.value_of("config").unwrap());
+            if let Some(config) = server_config::from_path(config_path) {
+                Command::Server(config)
+            } else {
+                bail!("Could not load config")
+            }
         },
-        _ => bail!("no subcommand specified"),
+        _ => bail!("No subcommand specified"),
     })
 }
 
+fn create_server_token(server_id: ServerId, auth_token: &str) -> String {
+    format!("{} {}", server_id.addr(), auth_token)
+}
+fn check_server_token(server_token: &str, auth_token: &str) -> Option<ServerId> {
+    let mut split = server_token.splitn(2, |c| c == ' ');
+    let server_addr = split.next().and_then(|addr| addr.parse().ok())?;
+    match split.next() {
+        Some(t) if t == auth_token => Some(ServerId(server_addr)),
+        Some(_) |
+        None => None,
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+struct ServerJwt {
+    server_id: ServerId,
+}
+fn create_jwt_server_token(server_id: ServerId, header: &jwt::Header, key: &[u8]) -> String {
+    jwt::encode(&header, &ServerJwt { server_id }, key).unwrap()
+}
+fn dangerous_unsafe_extract_jwt_server_token(server_token: &str) -> Option<ServerId> {
+    jwt::dangerous_unsafe_decode::<ServerJwt>(&server_token)
+        .map(|res| res.claims.server_id)
+        .ok()
+}
+fn check_jwt_server_token(server_token: &str, key: &[u8], validation: &jwt::Validation) -> Option<ServerId> {
+    jwt::decode::<ServerJwt>(server_token, key, validation)
+        .map(|res| res.claims.server_id)
+        .ok()
+}
+
 fn run(command: Command) -> Result<i32> {
     match command {
-        Command::Scheduler => {
+        Command::Auth(AuthSubcommand::Base64 { num_bytes }) => {
+            let mut bytes = vec![0; num_bytes];
+            let mut rng = rand::OsRng::new().unwrap();
+            rng.fill_bytes(&mut bytes);
+            // As long as it can be copied, it doesn't matter if this is base64 or hex etc
+            println!("{}", base64::encode_config(&bytes, base64::URL_SAFE_NO_PAD));
+            Ok(0)
+        },
+        Command::Auth(AuthSubcommand::JwtHS256ServerToken { secret_key, server_id }) => {
+            let header = jwt::Header::new(jwt::Algorithm::HS256);
+            let secret_key = base64::decode_config(&secret_key, base64::URL_SAFE_NO_PAD)?;
+            println!("{}", create_jwt_server_token(server_id, &header, &secret_key));
+            Ok(0)
+        },
+
+        Command::Scheduler(scheduler_config::Config { client_auth, server_auth }) => {
+            let check_client_auth: Box<Fn(&str) -> bool + Send + Sync> = match client_auth {
```

Comment:
Oops, I meant to put that comment *here*.

Reply:
I was just thinking that extracting the meat of this match arm into a separate function would make this function smaller and easier to follow. It's not a huge concern.